languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/SessionFactoryTest.java

```diff
@@ -42,32 +49,677 @@
+        .withInMemoryMetastore()
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      int numThreads = 100;
+      int numTasks = numThreads * 100;
+      ExecutorService pool = Executors.newFixedThreadPool(numThreads);
+      LongAdder tasksCompleted = new LongAdder();
+      IntStream.range(0, numTasks)
+        .parallel()
+        .forEach(i -> {
+            pool.submit(() -> {
+              try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+                byte[] payload = Integer.toHexString(i).getBytes();
+                byte[] drr = session.encrypt(payload);
+                byte[] decryptedPayload = session.decrypt(drr);
+
+                assertArrayEquals(payload, decryptedPayload);
+                tasksCompleted.increment();
+              } catch (Exception e) {
+                fail(e.getMessage());
+              }
+            });
+        });
+
+      try {
+        pool.shutdown();
+        pool.awaitTermination(60, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      assertEquals(numTasks, tasksCompleted.sum());
+      assertEquals(1, factory.getSessionCache().estimatedSize());
+    }
+  }
+
+  @Test
+  void testSessionCacheMultiThreadedWithMaxSessionReachedDifferentSessions() {
+    CryptoPolicy policy = BasicExpiringCryptoPolicy.newBuilder()
+        .withKeyExpirationDays(1)
+        .withRevokeCheckMinutes(30)
+        .withCanCacheSessions(true)
+        .withSessionCacheMaxSize(1)
+        .build();
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withInMemoryMetastore()
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      int numThreads = 100;
+      int numTasks = numThreads * 100;
+      ExecutorService pool = Executors.newFixedThreadPool(numThreads);
+      LongAdder tasksCompleted = new LongAdder();
+      IntStream.range(0, numTasks)
+        .parallel()
+        .forEach(i -> {
+            pool.submit(() -> {
+              try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId + i)) {
+                byte[] payload = Integer.toHexString(i).getBytes();
+                byte[] drr = session.encrypt(payload);
+                byte[] decryptedPayload = session.decrypt(drr);
+
+                assertArrayEquals(payload, decryptedPayload);
+                tasksCompleted.increment();
+              } catch (Exception e) {
+                fail(e.getMessage());
+              }
+            });
+        });
+
+      try {
+        pool.shutdown();
+        pool.awaitTermination(60, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      assertEquals(numTasks, tasksCompleted.sum());
+      assertEquals(1, factory.getSessionCache().estimatedSize());
```

Comment:
Instead of verifying this, we can verify after L:604 that number of entries in cache is less than the maxCacheSize defined.

Reply:
similar comment here on not being possible in this test case. had to remove this assertion anyway since we couldn't rely on the `estimatedSize` call (covered in comment in the test case)